rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,500 @@
+
+[optional-argument]: https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwAc4A3XZAZwAooAuecjGLVAcwBp4AjO1ZAW04gYASjrEcWYAG4AUKEiwEKdNjzxkBYFAwhg1OgyasOnAPw9+gkWInSZMiCAzwo8ALyESZKgHICODG0cHw4ARgAmYVlHZ053dU1tXUo-AKCQ+AiooA
+
+[narrower-property]: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEkoGdnwPIwJYHMsDsoJ4BvAKHngAcYB7KkGAFwE8AueZJ7fHeAH07cCOANoBdANxkAvmTKgkcRNFTwAcrDoB3EMFIVqdBs3ZCeOWfIBmAV3xgmWWvngALKPmAQQyTLgIiTRgdAAoANyJbEA5-PEIIAEoOfFsAWwAjRgNKOCZbGFdIiGiAOhp6RlZSn14mN2k5BXBoZR8meChY7HiiaUVWhHb4DI5gnT1pDy8fPx7AiHHabVCoRKnPb184haWVjPWyIA
+
+#### Bug fixes to type definitions
+
+As with runtime code, types may have bugs. Addon authors may find cases where previously-allowed code was allowed *incorrectly*. For example, a type may have been `any` accidentally, allowing users to instantiate classes or call functions or methods incorrectly. In these cases, the incorrect types will correspond to incorrect runtime behavior! Just as with bug fixes for runtime code, addon authors may publish fixes to *types* in patch releases.
+
+#### Dropping support for previously-supported versions
+
+Additionally, once a TypeScript version has been added to an addon’s list of supported versions, dropping it constitutes a breaking change, because it has the same kind of impact on users of the addon as dropping support for a version of Ember, Ember CLI, or Node. Whatever the reason for dropping a previously-supported TypeScript release, addons should publish a new major version.
+
+### Tooling
+
+To successfully use TypeScript, we need to be able to *detect* breaking changes (whether from our own changes or from TypeScript itself) and to *mitigate* them.
+
+#### Detect breaking changes in types
+
+As with runtime code, it is essential to prevent unintentional changes to the API of types supplied by an addon. We can accomplish this using *type tests*: tests which assert that the types exposed by the public API of the addon are stable.
+
+Addon authors publishing types should add [the `expect-type` library][expect-type] library and generate a set of tests corresponding to their public API. These tests should be put in the `types` folder configured by `ember-cli-typescript`’s default generator. If the API surface is small, a single file may be sufficient. For example, `ember-modifier` has `types/ember-modifier-test.ts`. In more complicated packages, addon authors may choose to break the tests down to mirror the structure of the addon’s API—for example, by dedicated test files for each helper, modifier, service, component, or utility class or function exported from the addon.
```

Comment:
What's the rationale behind choosing `expect-type` over `dtslint` and other options? Does this have a level of TS community adoption and support that we're comfortable with?

Reply:
After sleeping on this (the use of `expect-type` as opposed to `dtslint`, `tsd` or something else), I feel like this RFC should state _problem(s) that any chosen testing solution must/should/may solve and/or mitigate_. 

Evaluating one or more existing solutions that address this "problem statement" would be a lot easier, and we'd be able to more clearly evaluate whether _any_ of the pre-existing tools meet our needs.

Reply:
I've updated this section with a discussion of the constraints as well as the three major possible tools here and noted that while `expect-type` is currently the best option, but not making a formal recommendation here. We can come back to that later. Marking this as resolved at this point; let me know if there's anything I missed in the updated text!